probe-rs/src/probe/ftdi/ftdaye/d2xx.rs

```diff
@@ -0,0 +1,87 @@
+use std::time::Duration;
+
+use libftd2xx::{Ftdi, FtdiCommon};
+use nusb::DeviceInfo;
+
+use crate::DebugProbeError;
+
+use super::{error::FtdiError, FtdiDriver, Interface, Result};
+// todo https://docs.rs/libloading/latest/libloading/
+
+/// An FTDI driver using the proprietary D2XX driver installed by default in
+/// Windows. Should technically work in Linux, but untested
+pub struct FtdiD2xx {
+    ft: Ftdi,
+}
+
+fn ft_status_to_lib_err(e: libftd2xx::FtStatus) -> FtdiError {
+    FtdiError::Other(format!("FTDI D2XX error: {e}"))
+}
+
+fn ft_status_to_io_err(e: libftd2xx::FtStatus) -> std::io::Error {
+    std::io::Error::other(format!("FTDI D2XX error {e}"))
+}
+
+impl FtdiDriver for FtdiD2xx {
+    fn usb_reset(&mut self) -> Result<()> {
+        self.ft.reset().map_err(ft_status_to_lib_err)
+    }
+
+    fn usb_purge_buffers(&mut self) -> Result<()> {
+        self.ft.purge_all().map_err(ft_status_to_lib_err)
+    }
+
+    fn set_usb_timeouts(
+        &mut self,
+        read_timeout: std::time::Duration,
+        write_timeout: std::time::Duration,
+    ) -> Result<()> {
+        self.ft
+            .set_timeouts(read_timeout, write_timeout)
+            .map_err(ft_status_to_lib_err)
+    }
+
+    fn set_latency_timer(&mut self, value: u8) -> Result<()> {
+        self.ft
+            .set_latency_timer(Duration::from_millis(value as u64))
+            .map_err(ft_status_to_lib_err)
+    }
+
+    fn set_bitmode(&mut self, bitmask: u8, mode: super::BitMode) -> Result<()> {
+        self.ft
+            .set_bit_mode(bitmask, (mode as u8).into())
+            .map_err(ft_status_to_lib_err)
+    }
+
+    fn read_data(&mut self, data: &mut [u8]) -> std::io::Result<usize> {
+        self.ft.read(data).map_err(ft_status_to_io_err)
+    }
+
+    fn write_data(&mut self, data: &[u8]) -> std::io::Result<usize> {
+        self.ft.write(data).map_err(ft_status_to_io_err)
+    }
+}
+
+impl FtdiD2xx {
+    pub fn open(usb_device: &DeviceInfo, interface: Interface) -> Result<Self, DebugProbeError> {
+        if interface != Interface::A {
+            return Err(DebugProbeError::NotImplemented(
+                "Non-default FTDI interfaces",
+            ));
+        }
+        let Some(serial) = usb_device.serial_number() else {

```

Comment:
I have multiple FTDI devices without external EEPROMs, and thus without serial numbers.

Reply:
As far as I can tell, you will still have a serial number, it will just be the empty string, so I think it'll still work in your case.

Not being able to map the nusb device path to ftdi device path certainly sucks.

Reply:
Unfortunately this definitely returns an error for me with FTDIs without external EEPROMs.

Reply:
Blast, thanks for testing. Do you often need to target a particular device? Could this just lazily use the first device instance? Or should I rework it and try another identification scheme